SIMSBIOHUB-579/583: Updated Animal Workflow (Capture, Mortality, Measurements, Profile)

mauberti-bc opened the pull request:
## Links to Jira Tickets

- https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-579
- https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-583

## Description of Changes

Refactors the Manage Animals-related pages to follow SIMS design patterns

- Create/edit/delete animal workflow
- Create/edit/delete capture workflow
- Create/edit/delete mortality workflow
- Animal profile to display above info

## Testing Notes

Requires https://github.com/bcgov/critterbase-api/pull/54, which fixes several related bugs in Critterbase endpoints.

mauberti-bc committed:
make install

mauberti-bc committed:
animal profile page & context

mauberti-bc committed:
Merge branch 'dev' of github.com:bcgov/biohubbc into manage-animals

mauberti-bc committed:
wip: animal profile page

mauberti-bc committed:
wip: add capture event

mauberti-bc committed:
release location

mauberti-bc committed:
edit animal

mauberti-bc committed:
update context to set selectedanimal from survey critter id

mauberti-bc committed:
animal list

mauberti-bc committed:
wip: report mortality

mauberti-bc committed:
edit animal page loading spinner

mauberti-bc committed:
no captures placeholder

mauberti-bc committed:
fix objectives in openapi project update request

mauberti-bc committed:
update edit capture page

mauberti-bc committed:
wip: formik validation

mauberti-bc committed:
initial formik validation on capture form

mauberti-bc committed:
jsdocs

mauberti-bc committed:
Merge branch 'dev' into manage-animals

mauberti-bc committed:
animal profile improvements

mauberti-bc committed:
profile

mauberti-bc committed:
merge latest

mauberti-bc committed:
wip: collection units

mauberti-bc committed:
ecological units

mauberti-bc committed:
update population units in edit critter

mauberti-bc committed:
wip: markings

mauberti-bc committed:
eco units

mauberti-bc committed:
wip: capture form

mauberti-bc committed:
merge latest

mauberti-bc committed:
merge in dev

mauberti-bc committed:
broken: ecological units data loader

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into manage-animals

NickPhura committed:
Get Animal form control interaction working.
Update some related interfaces to match response from critterbase.

mauberti-bc committed:
make animal profile components dumber

mauberti-bc committed:
wip: delete critters from survey

NickPhura committed:
Add isEdit mode support for create/edit animal page

mauberti-bc committed:
markings card on capture form

mauberti-bc committed:
Merge branch 'manage-animals' of github.com:bcgov/biohubbc into manage-animals

mauberti-bc committed:
wip: adding markings to captures

mauberti-bc committed:
delete markings linked to capture before deleting capture

NickPhura committed:
Update animal components

NickPhura committed:
Merge branch 'manage-animals' of https://github.com/bcgov/biohubbc into manage-animals

NickPhura committed:
Tweaks

mauberti-bc committed:
add and edit measurements on capture page

mauberti-bc committed:
wip: fixing measurement types

mauberti-bc committed:
add and edit markings on capture

mauberti-bc committed:
fix marking update

mauberti-bc committed:
merge manage-animals branch

mauberti-bc committed:
initial report mortality workflow

mauberti-bc committed:
mortality section of animal profile

mauberti-bc committed:
add measurements and markings to capture card on animal profile

mauberti-bc committed:
fix app index file

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into manage-animals

NickPhura committed:
Minor cleanup and organization changes. Wrap some dataloaders in useEffects, etc.

NickPhura committed:
Wrap load() calls in useEffect

NickPhura committed:
Fix edit capture requests (was sending measurements/markings from other capture events).

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into manage-animals

NickPhura committed:
Cleanup, Fix required marking identifier

NickPhura committed:
Merge remote-tracking branch 'origin/manage-animals' into animal-mortality

NickPhura committed:
Merge in latest animal-capture branch, fix merge issues.

NickPhura committed:
Misc

NickPhura committed:
Fix

NickPhura committed:
WIP

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into animal-mortality

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into animal-mortality

NickPhura committed:
Updates

NickPhura committed:
Update mortality CRUD

mauberti-bc committed:
fix capture_release data in critterbase payload & styling

NickPhura committed:
Remove datagrid hover css.
Fix useEffect dependency warning

MacDeluca commented on app/src/components/species/components/SpeciesAutocompleteField.tsx:
```diff
@@ -141,7 +141,7 @@ const SpeciesAutocompleteField = (props: ISpeciesAutocompleteFieldProps) => {
         if (option) {
           handleSpecies(option);
-          setInputValue(startCase(option.commonNames[0] ?? option.scientificName));
+          setInputValue(startCase(option?.commonNames ? option.commonNames[0] : option.scientificName));
```
an empty array will evaluate to true
option?.commonNames.length ? option.commonNames[0] : option.scientificName

NickPhura replied:
Updated

MacDeluca commented on app/src/constants/spatial.ts:
```diff
@@ -21,9 +21,9 @@ export const ALL_OF_BC_BOUNDARY: Feature = {
 
 export const SURVEY_MAP_LAYER_COLOURS = {
-  OBSERVATIONS_COLOUR: '#db4f4f',
+  OBSERVATIONS_COLOUR: '#1f6fb5', //'#db4f4f',
```
👀  Previous colour?

NickPhura replied:
Removed

MacDeluca commented on app/src/features/surveys/animals/AnimalPage.tsx:
```diff
@@ -0,0 +1,79 @@
+    biohubApi.survey.getSurveyCritters(surveyContext.projectId, surveyContext.surveyId)
+  );
+
+  useEffect(() => {
```
Is the useEffect needed here?

NickPhura replied:
I believe so, otherwise you get warnings about components updating while other components are rendering.

MacDeluca commented on app/src/features/surveys/animals/animal-form/edit/EditAnimalPage.tsx:
```diff
@@ -0,0 +1,243 @@
+        itis_tsn: values.species.tsn,
+        wlh_id: values.wildlife_health_id,
+        animal_id: values.nickname,
+        sex: AnimalSex.UNKNOWN,
```
Should this be set to unknown?

NickPhura replied:
I don't think there is a sex field on the new create animal form, though maybe there is meant to be and this was a placeholder that got missed? 

Probably a @mauberti-bc question.

mauberti-bc replied:
I hardcoded sex to unknown in the critter table so that users add sex as a measurement instead, which gives users more relevant options for their species. We can keep it as is.

NickPhura replied:
Do we need to deprecate that sex field in Critterbase then?

We'd want to eventually migrate the existing records to using a measurement then too.

mauberti-bc replied:
Yeah, eventually we will

MacDeluca commented on app/src/features/surveys/animals/list/AnimalListContainer.tsx:
Is this component needed?

NickPhura replied:
I moved the code from `AnimalList` to `AnimalListContainer` since it was basically acting as the parent component already, and then removed `AnimalList`

MacDeluca commented on app/src/features/surveys/animals/profile/captures/capture-form/components/AnimalCaptureForm.tsx:
```diff
@@ -0,0 +1,118 @@
+    }),
+    measurements: yup.array(
+      yup.object({
+        // TODO
```
👀

MacDeluca commented on app/src/features/surveys/animals/profile/captures/components/capture-card-details/components/CaptureDetails.tsx:
```diff
@@ -0,0 +1,79 @@
+ * @param {ICaptureDetailsProps} props
+ * @return {*}
+ */
+export const CaptureDetails = (props: ICaptureDetailsProps) => {
```
Do we need CaptureDetails and ReleaseDetails components?
Could these be condensed into one component that accepts a generic object?

NickPhura replied:
Maybe, though I don't think in a nice way. The capture and release types have different nullable vs non-nullable fields, and come from different fields in the same response object, so we would have to pre-map them to generic props of the generic component, which just adds more complexity I feel.

MacDeluca commented on app/src/features/surveys/animals/profile/components/MarkingDetails.tsx:
```diff
@@ -0,0 +1,45 @@
+      </Typography>
+      <Box maxHeight="600px" sx={{ overflowY: 'auto', pr: 1 }}>
+        {markings.map((marking, index) => (
+          <Box mt={1} key={`${marking.marking_id}-${index}`}>
```
Is this a spot where we need to use index?

NickPhura replied:
Doesn't look like it. Will remove.

I know I sometimes get tripped up when it comes to knowing which ids are unique (primary keys) and which are code value ids and therefore not unique (like the id of an ear tag marking, which a critter might have 2 of).

MacDeluca commented on app/src/features/surveys/animals/profile/components/MeasurementDetails.tsx:
```diff
@@ -0,0 +1,45 @@
+    return null;
+  }
+
+  const allMeasurements = [...measurements.quantitative, ...measurements.qualitative];
```
useMemo?

MacDeluca replied:
This is one of those places where I am uncertain if useMemo is more performant or less

MacDeluca replied:
Ignore.

MacDeluca commented on app/src/features/surveys/animals/profile/components/MeasurementDetails.tsx:
```diff
@@ -0,0 +1,45 @@
+          <Paper
+            variant="outlined"
+            sx={{ px: 3, py: 2, bgcolor: grey[100], mt: 1 }}
+            key={`${measurement.taxon_measurement_id}-${index}`}>
```
Do we need index?
Or should be generate indexes for these values ie: _id: uuid()

NickPhura replied:
Not sure? I can change this to just generating a uuid

NickPhura replied:
It probably doesn't matter too much in this case, as the list of measurements is static. There are no UI controls to manipulate it (they are on a different page entirely).

MacDeluca commented on app/src/features/surveys/animals/profile/mortality/AnimalMortalityContainer.tsx:
```diff
@@ -0,0 +1,118 @@
+    return null;
+  }
+
+  const mortality: IMortalityWithSupplementaryData[] =
```
useMemo

NickPhura replied:
Update

MacDeluca commented on app/src/features/surveys/animals/profile/mortality/AnimalMortalityContainer.tsx:
```diff
@@ -0,0 +1,118 @@
+ *
+ * @return {*}
+ */
+const AnimalMortalityContainer = () => {
```
Does this component need useUnsavedChangesDialog()?
I just noticed we are using a history.push() onSave

NickPhura replied:
Not needed in this component. This particular push is just navigating to the create mortality page (not a callback on saving a mortality).

NickPhura replied:
That being said, its being passed in as a prop, but probably it could just be in the child component directly. Probably just trying to make the child component simpler/dumber, which isn't a bad thing.

MacDeluca commented on app/src/features/surveys/animals/profile/mortality/mortality-form/components/cause-of-death/CauseOfDeathForm.tsx:
```diff
@@ -0,0 +1,44 @@
+import { ICreateEditMortalityRequest } from 'interfaces/useCritterApi.interface';
+import { useEffect } from 'react';
+
+export const CauseOfDeathForm = () => {
```
jsdoc

NickPhura replied:
Added

MacDeluca commented on app/src/features/surveys/animals/profile/mortality/mortality-form/components/general-information/MortalityGeneralInformationForm.tsx:
```diff
@@ -0,0 +1,51 @@
+import { useFormikContext } from 'formik';
+import { ICreateEditMortalityRequest } from 'interfaces/useCritterApi.interface';
+
+export const MortalityGeneralInformationForm = () => {
```
jsdoc

NickPhura replied:
Added

MacDeluca commented on app/src/features/surveys/animals/profile/mortality/mortality-form/components/location/MortalityLocationForm.tsx:
```diff
@@ -0,0 +1,15 @@
+import Box from '@mui/system/Box';
+import { MortalityLocationMapControl } from './MortalityLocationMapControl';
+
+export const MortalityLocationForm = () => {
```
jsdoc

NickPhura replied:
Added

MacDeluca commented on app/src/features/surveys/animals/profile/mortality/mortality-form/create/CreateMortalityPage.tsx:
```diff
@@ -0,0 +1,274 @@
+  measurements: []
+};
+
+const CreateMortalityPage = () => {
```
jsdoc

NickPhura replied:
Added

MacDeluca commented on app/src/features/surveys/observations/sampling-sites/list/SamplingSiteList.tsx:
```diff
@@ -339,6 +339,7 @@ const SamplingSiteList = () => {
                 <Divider flexItem></Divider>
                 <Box
                   flex="1 1 auto"
+                  // overflow="hidden"
```
🥽

NickPhura replied:
Removed

MacDeluca commented on app/src/hooks/cb_api/useXrefApi.tsx:
```diff
@@ -29,8 +31,32 @@ export const useXrefApi = (axios: AxiosInstance) => {
+   * @param {number} tsn
+   * @return {*}  {Promise<ICollectionCategory[]>}
+   */
+  const getCollectionCategories = async (tsn: number): Promise<ICollectionCategory[]> => {
```
maybe rename to getTsnCollectionCategories?

MacDeluca replied:
I think renaming this makes sense if we would ever have a frontend hook that would retrieve all collection categories

NickPhura replied:
Renamed

NickPhura replied:
Happy to change the name to `getTsnCollectionCategories` if you want though.

Made it feel like the collection categories belonged to the TSN, which is maybe true, but they are their own thing, just filtered by TSN. Dunno.

MacDeluca commented on app/src/hooks/cb_api/useMarkingApi.tsx:
```diff
@@ -1,8 +1,47 @@
+   * @param {number} tsn
+   * @returns {Promise<IMarkingTypeResponse>} - The created marking.
+   */
+  const getMarkingBodyLocationOptions = async (tsn: number): Promise<IMarkingBodyLocationResponse[]> => {
```
maybe rename to getTsnMarkingBodyLocationOptions?

NickPhura replied:
Renamed

MacDeluca commented:
Few things I noticed:

1. Looks like the markings cards are missing some margins between them (ignore if intended)
![image](https://github.com/bcgov/biohubbc/assets/99926243/0d526b93-d3a3-43dd-a124-e76a131be3b4)
2. The save on Mortality failed when entered a measurement value (I cleared and re-added). Looks like the value is not being removed when it switches from qualitative<->quantitative
error: 
```json

{
    "error": "Zod validation failed.",
    "issues": [
        {
            "code": "unrecognized_keys",
            "keys": [
                "value"
            ],
            "path": [
                0
            ],
            "message": "Unrecognized key(s) in object: 'value'"
        }
    ]
}
```
request body: 
```json

{
    "markings": [
        {
            "_id": "cab39c06-062a-4419-968c-69b8f03f30c0",
            "marking_type_id": "3e24234e-dea6-445d-9e18-5987fad81d67",
            "taxon_marking_body_location_id": "3618a897-fee3-45d3-8629-6e1c6a6c5ec4",
            "identifier": "asdfasdf",
            "primary_colour_id": "618c811b-4e05-462a-a01f-d3814613ae79",
            "secondary_colour_id": "618c811b-4e05-462a-a01f-d3814613ae79",
            "comment": "asdfasdfasdf",
            "critter_id": "8fc1f48a-4328-43d3-aa07-da0684e89857",
            "mortality_id": "1cb8468b-d32e-40f7-b561-20e1e6a2524c"
        },
        {
            "_id": "272974c6-137f-4c23-a344-4cb102bab78e",
            "marking_type_id": "3e24234e-dea6-445d-9e18-5987fad81d67",
            "taxon_marking_body_location_id": "1c028dcf-9108-49b7-9a51-be233cbf0c8b",
            "identifier": "asdfasdf",
            "primary_colour_id": "0a861e69-e452-4a8c-a3bd-6ddf362254bc",
            "secondary_colour_id": null,
            "comment": null,
            "critter_id": "8fc1f48a-4328-43d3-aa07-da0684e89857",
            "mortality_id": "1cb8468b-d32e-40f7-b561-20e1e6a2524c"
        }
    ],
    "qualitative_measurements": [
        {
            "taxon_measurement_id": "30d6ca71-e4a7-419b-b392-225332f4db44",
            "qualitative_option_id": "8646875f-34d8-4469-9500-6a7cd30d88ae",
            "value": 234,
            "critter_id": "8fc1f48a-4328-43d3-aa07-da0684e89857",
            "mortality_id": "1cb8468b-d32e-40f7-b561-20e1e6a2524c"
        }
    ],
    "quantitative_measurements": [
        {
            "taxon_measurement_id": "30d6ca71-e4a7-419b-b392-225332f4db44",
            "qualitative_option_id": "8646875f-34d8-4469-9500-6a7cd30d88ae",
            "value": 234,
            "critter_id": "8fc1f48a-4328-43d3-aa07-da0684e89857",
            "mortality_id": "1cb8468b-d32e-40f7-b561-20e1e6a2524c"
        }
    ]
}
```

MacDeluca committed:
docs: updated env.docker to include linux container hosts

NickPhura commented:
1. Looks like the markings cards are missing some margins between them (ignore if intended)

- Fixed

2. The save on Mortality failed when entered a measurement value (I cleared and re-added). Looks like the value is not being removed when it switches from qualitative<->quantitative

- Fixed

NickPhura committed:
Minor fixes

NickPhura committed:
Updates

NickPhura committed:
Updates

NickPhura committed:
Fix/improve capture/mortality create/edit measurements

NickPhura commented:
@MacQSL Okay I've pushed all of my review related changes up. I believe I've covered everything. Ill give it one more round of manual tests tomorrow morning to make sure.

NickPhura committed:
Fix code smells

NickPhura committed:
Fix test

NickPhura committed:
Fix collapse UI behaviour. Minor Tweaks.

NickPhura committed:
Cleanup

NickPhura committed:
Fix typings

MacDeluca commented:
Few minor tech debt things which are outside scope of this PR, but putting here to track them.

1. The capture map icons appear clickable, maybe we should disable that or allow the capture to be edited on click?
![image](https://github.com/bcgov/biohubbc/assets/99926243/08373acf-dbfa-4401-84c1-ea0aef4bb77c)
2. The icon beside 'Cervidae' almost appears to be a tooltip or some hoverable info action.
![image](https://github.com/bcgov/biohubbc/assets/99926243/15ad5609-83da-477b-aeeb-b09d9f0f6952)
3. The release locations don't appear on the map (maybe they should?)
4. The measurement cards look like they could use some additional styling help
![image](https://github.com/bcgov/biohubbc/assets/99926243/ab5fe83a-8e39-440a-b36e-b817e02accc9)
5. Feels like the capture / mortality map components should share the same map component?

MacDeluca commented:
The edit animal action from the 'Survey Page' doesn't select the animal in the manage animals page. 
The previous implementation was using url query params to keep the selected animal state. This will also fix the animal from being unselected on a page refresh.
![image](https://github.com/bcgov/biohubbc/assets/99926243/dca17b38-988c-4427-bd6b-86c058b47f3b)

NickPhura committed:
Add start case to measurement/marking labels.
Add common names, etc, to animal card when editing existing animal.
Fix route from survey animal table when clicking action menu 'edit'.

NickPhura commented:
> The edit animal action from the 'Survey Page' doesn't select the animal in the manage animals page. The previous implementation was using url query params to keep the selected animal state. This will also fix the animal from being unselected on a page refresh. ![image](https://private-user-images.githubusercontent.com/99926243/338780681-dca17b38-988c-4427-bd6b-86c058b47f3b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxNTE3MTIsIm5iZiI6MTcxODE1MTQxMiwicGF0aCI6Ii85OTkyNjI0My8zMzg3ODA2ODEtZGNhMTdiMzgtOTg4Yy00NDI3LWJkNmItODZjMDU4YjQ3ZjNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDAwMTY1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxYjM4MTEwZWNhOGUyOWI1NjM5ODc5Zjc2NWE1YzE5MjI0YjE1YTUyMjQ1NjIxZDIzNzFjMjRhNjIxYjUzYjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GMrebCb5JCaC8G3_0FZTqyInECzMsLdE7qir5aOE_2U)

This should be fixed.

I also added 'startCase' calls to the marking and measurement labels, and updated the species card to show common names, etc, when editing an existing animal.

MacDeluca reviewed (approved):
🤯
